src/Context/product-context.js

```diff
@@ -0,0 +1,637 @@
+                        }
+                    }
+                }
+                return [...state]
+            }
+
+        case "UPDATE_LIST_AS_PER_FILTERS" : 
+            {
+                if(action.payload.includeOutOfStock===true)
+                {
+                    //All products
+                    return (action.payload.onlyFastDeliveryProducts===false) 
+                    ? productList.filter(item=>
+                        (
+                            action.payload.minPrice<=item.discountedprice 
+                            && item.discountedprice<action.payload.maxPrice
+                            && action.payload[item.genre]
+                            && item.rating>=action.payload.minRating
+                        )
+                      )
+                    : productList.filter(item=> 
+                        (
+                            item.fastDeliveryAvailable===true
+                            && (
+                                action.payload.minPrice<=item.discountedprice 
+                                && item.discountedprice<action.payload.maxPrice
+                               ) 
+                            && action.payload[item.genre]
+                            && item.rating>=action.payload.minRating
+                        ))
+                }
+                else
+                {
+                    //Only available products
+                    return (action.payload.onlyFastDeliveryProducts===false) 
+                    ? productList.filter(item=>
+                        (
+                            item.outOfStock===false 
+                            && (
+                                action.payload.minPrice<=item.discountedprice 
+                                && item.discountedprice<action.payload.maxPrice
+                               ) 
+                            && action.payload[item.genre]
+                            && item.rating>=action.payload.minRating
+                        ))
+                    : productList.filter(item=>
+                        (
+                            item.outOfStock===false
+                            && item.fastDeliveryAvailable===true 
+                            && (
+                                action.payload.minPrice<=item.discountedprice 
+                                && item.discountedprice<action.payload.maxPrice
+                               ) 
+                            && action.payload[item.genre]
+                            && item.rating>=action.payload.minRating
+                        ))
+                }
+            }
+
+        default : return [...state]
+    }
+}
+
+let ProductsProvider = ({children}) => 
+{
+    const [ productsAvailableList, dispatchSortedProductsList] = useReducer(productsOrderFunc,productList)
+    const [ productFilterOptions, dispatchProductFilterOptions ] = useReducer(updateProductFilters,filterOptionsObject)
+
```

Comment:
Suggestion: I think this violates the rules of hooks read this https://reactjs.org/docs/hooks-rules.html

Reply:
I believe no rules were broken, I called useReducer at the top level of a component. Would be great if you can elaborate a little more 😅